### PR TITLE
Fix while loop parsing in in protobuf format

### PR DESCRIPTION
**Pull request by shraiysh**
While loop computations were not being linked to the while operation. This patch fixes that and connects them.

### CODE REVIEW

**Review by frgossen — CHANGES_REQUESTED** [review]
Can you add a test case?

**Comment by shraiysh** [comment]
Do you know where testcases for this would be added?

**Comment by frgossen** [comment]
> Do you know where testcases for this would be added?

hlo_parser_test.cc

**Comment by shraiysh** [comment]
@frgossen modified tests to check for this too.

**Review by frgossen — CHANGES_REQUESTED** [review]
I looked a bit more and think the parser should actually not populate this because it is ambiguous.

Example: For a computation that is used as a body for two distinct while ops, there is no way to populate this field correctly. With this change, you will just override it and keep the last one which is at least misleading. This field can only be populated correctly after flattening the call graph.

This field is populated in `xla/service/flatten_call_graph.cc` together with other caller/callee pointers.

File: xla/service/hlo_parser_test.cc [hunk 1]
```diff
@@ -2541,6 +2541,14 @@ class HloParameterizedParserTest
           original,
           module->ToString(HloPrintOptions().set_print_large_constants(true)));
     }
+    for (HloComputation* computation : module->computations()) {
```
Comment (frgossen) [inline xla/service/hlo_parser_test.cc:2544]: I don't think this is the right place to add an assertion. 
Can you add a test below to make sure that the field is null after parsing?

**Review by frgossen — CHANGES_REQUESTED** [review]

**Comment by shraiysh** [comment]
I understand the concern, however, this is also done while [parsing HLO as text](https://github.com/openxla/xla/blob/15a888968d354c92b34d0d04cd962e8883ab4275/xla/service/hlo_parser.cc#L2155). The CreateWhileOp call sets this. I think we should fix that too in this case. Another thing is that some execution (for example, [address computation fusion here](https://github.com/openxla/xla/blob/15a888968d354c92b34d0d04cd962e8883ab4275/xla/service/gpu/fusions/custom.cc#L179) depend on this being already set up. Does this mean that we should run call graph flattening everytime we parse HLO even if it is already optimized?

**Comment by frgossen** [comment]
@akuegel , you approved https://github.com/openxla/xla/pull/7179. Do you know what is the right thing to do here? 
- Populate the field and be aware that it could be ambiguous 
- Not populate it

**Comment by akuegel** [comment]
> @akuegel , you approved #7179. Do you know what is the right thing to do here?
> 
> * Populate the field and be aware that it could be ambiguous
> * Not populate it

Yes, I approved it, and also suggested to use a back pointer. I was not aware of the difficulties with adding back pointers. We have an internal bug open for this. Not only while is affected, other ops with back pointers are potentially affected as well. Just for fusions we are on the safe side, because it is an op internal to XLA, so we control that there is always a 1:1 relationship between fusion instruction and fusion computation.

**Review by frgossen — APPROVED** [review]
Thanks@
